src/librustc_codegen_llvm/back/write.rs

```diff
@@ -389,33 +488,17 @@ pub(crate) unsafe fn optimize(
                     llvm::LLVMPassManagerBuilderPopulateModulePassManager(b, mpm);
                 });
 
-                have_name_anon_globals_pass = have_name_anon_globals_pass || prepare_for_thin_lto;
                 if using_thin_buffers && !prepare_for_thin_lto {
                     llvm::LLVMRustAddPass(mpm, find_pass("name-anon-globals").unwrap());
-                    have_name_anon_globals_pass = true;
                 }
             } else {
                 // If we don't use the standard pipeline, directly populate the MPM
                 // with the extra passes.
                 for pass in extra_passes {
                     llvm::LLVMRustAddPass(mpm, pass);
                 }
-            }
-
-            if using_thin_buffers && !have_name_anon_globals_pass {
```

Comment:
Hmm, this was originally added because we did want `no_prepopulate_passes` to do what it says on the box.

+ at this point if `using_thin_buffers && !no_prepopulate_passes` the pass should already have been added by code above (its a compiler bug otherwise) so I don’t see much justification of making this change.

Reply:
The reason for this change is that '-C passes` is not (and cannot be in its current form) supported by the new pass manager, so there is no way to explicitly enable the name-anon-globals pass in that case.

Right now, if you use `-C no-prepopulate-passes` (and ThinLTO buffers are used), you will always get an error if you don't also specify `-C passes=name-anon-globals`. We do not permit running without that pass. This makes the addition of `name-anon-globals` automatic, as it is the only thing that can be done anyway.

Reply:
`-Cno-prepopulate-passes` is predominantly a debugging mechanism and IIRC the ways to obtain LLVM IR (`--emit=llvm-ir`) *also* happen before any LTO is run.

Making it *not* do what it is supposed to do is… suboptimal, I feel. Given that `-Cpasses` is non-operable with new PM I don’t object to removal of this code, but… I’d still like to see at least *some* warning when `-Cno-prepopulate-passes` is not exactly doing what it says it does. We could do that even by establishing some sort of a conflicting relationship between this flag and `-Clto=thin` for instance.

Reply:
Assuming I'm not misremembering things, `-C lto=thin` isn't really the (main) problem here: We always make use of ThinLTO buffers (and using `name-anon-globals` is a requirement of those, not just ThinLTO itself) whenever we create or embed LLVM bitcode. This means that even if you disable ThinLTO, something like `--emit=llvm-bc` still requires `name-anon-globals` to run.

We could avoid using ThinLTO buffers for that and thus also avoid the need for `name-anon-globals`, but I'm not sure what fallout that would have (while `--emit=llvm-bc` is probably not used for that purpose, other cases that embed bitcode probably always need to be ThinLTO compatible.)

Reply:
Right, so how about just printing out a warning when we add the `name-anon-globals` pass in anyway? Its on the user then to figure out how to entirely disable any kind of LTO.